c++/src/kj/compat/http.c++

```diff
@@ -5978,16 +5978,35 @@ public:
     auto forked = paf.promise.fork();
 
     // 3. Create the ConnectResponseImpl
-    auto tunnel = kj::refcounted<ConnectResponseImpl>(kj::mv(paf.fulfiller),
-                                                      kj::mv(pipe.ends[0]));
+    auto response = kj::refcounted<ConnectResponseImpl>(kj::mv(paf.fulfiller),
+                                                        kj::mv(pipe.ends[0]));
 
     // 5. Call service.connect, passing in the tunnel.
     //    The call to tunnel->getConnectStream() returns a guarded stream that will buffer
     //    writes until the status is indicated by calling accept/reject.
-    auto connectStream = tunnel->getConnectStream();
-    auto promise = service.connect(hostCopy, *headersCopy, *connectStream, *tunnel)
-        .attach(kj::mv(hostCopy), kj::mv(headersCopy), kj::mv(tunnel), kj::mv(connectStream));
-
+    auto connectStream = response->getConnectStream();
+    auto promise = service.connect(hostCopy, *headersCopy, *connectStream, *response)
+        .eagerlyEvaluate([&response=*response](kj::Exception&& ex) {
+      // A few things need to happen here.
+      //   1. We'll log the exception.
+      //   2. We'll break the pipe.
+      //   3. We'll reject the status promise if it is still pending.
+      //
+      // We'll do all of this within the ConnectResponseImpl, however, since it
+      // maintains the state necessary here.
+      response.handleException(kj::mv(ex));
+    }).attach(kj::mv(hostCopy), kj::mv(headersCopy), kj::mv(response), kj::mv(connectStream));
```

Comment:
Note that lazy evaluation applies to destroying these attachments, too. They won't be destroyed until the pipe end that the promise is attached to is destroyed.

One thing you could do is, instead of using `.attach()`, actually capture all these values by move into the `eagerlyEvaluate()` callback. That way they will all be destroyed after that callback returns (or if it's canceled). It also conveniently avoids allocation of an extra `AttachmentPromiseNode`.

Reply:
oh, right, good point... and good idea.